### PR TITLE
Support timestamp forks and implement shanghaiTime

### PR DESCRIPTION
**Pull request by gezero**
Implement shanghaiTime including TimestampSchedule and associated infrastructure code. 
TimestampSchedule sits alongside the pre and post MergeProtocolSchedules in TransitionProtocolSchedule. 

General call pattern followed is that if a given timestamp precedes the first timestamp in the schedule, i.e. a pre-shanghai block, then delegate to the appropriate MergeProtocolSchedule.

cancunTime has also been implemented in order to effectively test fork order logic.

Note, some issues will follow on from this:
https://github.com/hyperledger/besu/issues/4793
https://github.com/hyperledger/besu/issues/4788
https://github.com/hyperledger/besu/issues/4789

### CODE REVIEW

**Comment by bobsummerwill** [comment]
Is there some EIP specification which is being implemented here?

I am having an issue finding a paper-trail (and prior discussions and justifications) for switching to date-based activation.

**Comment by siladu** [comment]
> Is there some EIP specification which is being implemented here?
> 
> I am having an issue finding a paper-trail (and prior discussions and justifications) for switching to date-based activation.

@bobsummerwill I've not seen anything written down other than on calls/discord. It was mentioned in the last 15 mins of the recent ACD151: Marius from geth team is going to implement an EIP for this. I think most/all client teams have implemented this in some form already though for withdrawals devnet interop. 
I think the justification is the ability to couple CL and EL forks, i.e. Capella and Shanghai should fork at the same time. CL uses forks based on Epochs, which are a proxy for timestamps.

There's also some hive tests PRs for Withdrawals that use timestamp activation: https://github.com/ethereum/hive/pull/659
https://notes.ethereum.org/@marioevz/Withdrawals-hive-testing

**Comment by ajsutton** [comment]
Yep, ultimately this will just be specified by the Shanghai EIP/meta-EIP though I'm not entirely sure if we're doing meta-EIPs for upgrades anymore (we might, I just lost track of the process).  Since the format of chain config files aren't standardised it basically just boils down to the fact that Shanghai will state "for any block where timestamp > X" rather than "for any block where number is > X".

Marius Van Der Wijden is putting together an EIP around this though (as per ACD notes) - I think that's predominantly for how the fork ID is calculated now but may cover more than that.

And as Simon notes it is a requirement for making withdrawals work that the CL and EL trigger the upgrade at the same time and that can't be done with block number because the CL needs to perform the transition at an epoch boundary and block number and slot number aren't tied together (can have empty slots that don't increase the block number).

**Comment by bobsummerwill** [comment]
Thanks for the clarification, @ajsutton.

Meta EIPs are no more, so it would make sense to capture this in an EIP, yes.

Best wishes!

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java
Comment (jframe) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java:24]: Is this class even needed? It's implementation is the same as `ProtocolSpecAdapters`.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java
Comment (jframe) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java:34]: There is a lot of duplication between this class and the ProtocolScheduleBuilder. Is it possible to reuse some of the code from `ProtocolScheduleBuilder`? Or just reuse the `ProtocolScheduleBuilder` passing the timestamp specific pieces?

File: besu/src/main/java/org/hyperledger/besu/controller/TransitionBesuControllerBuilder.java
Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/controller/TransitionBesuControllerBuilder.java:188]: ahah, I had this as a TODO but lost it in a stash. 

I was wondering if we should just wire it into all ProtocolSchedules, but there's a bit of circular dependency. Constructing ProtocolContext is dependent on a number of things that depend on ProtocolSchedule. 

TransitionProtocolSchedule previously used ProtocolContext and obtained it via a method argument in what is now called `getByBlockHeaderFromTransitionUtils`. Unfortunately, this method is abstracted inside getBlockByBlockHeader, an overridden method, so not trivial to add an argument to. 
This PR also uses ProtocolContext a second time, in `getByBlockNumber` which is an overridden method too. Hopefully this second usage will go away when we play https://github.com/hyperledger/besu/issues/4789 but the former usage might still be a problem.

I can't see an easy way to do this right now, would you be happy for it to be revisited as part of #4789?

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionUtils.java
Comment (siladu) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionUtils.java:106]: good spot! This was another lost TODO, I was thinking it could be wired into TPS but if that already has access, even better :)

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java:24]: Maybe not, think you're right. Probably could just rename ProtocolSpecAdapters' variable to work for both blockNumber and timestamp. Either way, this class will disappear when we play https://github.com/hyperledger/besu/issues/4788

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java:34]: I'll have a think about this. Some of the duplication goes away when we remove the modifiers as part of #4788.

There's enough slight differences that I'm worried it could make it unnecessarily complicated to understand the code.
I don't expect this code is going to change much either.
Also, eventually we'll probably only use the TimestampSchedule for public networks at least and ProtocolSchedule for private use case so they may naturally live separately in the future.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java:27]: I don't think we should exclude private use cases (I renamed MainnetTimestampSchedule -> DefaultTimestampSchedule to reflect this).

I didn't spot that anything was missed compared to the others, so I'll look into adding these, thanks!

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionProtocolSchedule.java
Comment (siladu) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionProtocolSchedule.java:131]: I agree, but after implementing the error here a number of tests started failing and I realised that these failures were not guarded by shanghaiTimestamp. I tried to resolve the failing tests and it opened a rabbit hole of pain that was essentially equivalent to the full implementation of https://github.com/hyperledger/besu/issues/4789 which I didn't want to bloat this PR with. 

I thought it was better to leave this known bug in (which *is* guarded by shanghaiTimestamp) and play #4789 ASAP

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolSpecBuilder.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolSpecBuilder.java:307]: I think it was guarding the block process as well, but the diff makes it hard to tell. This is what's on main: https://github.com/hyperledger/besu/blob/f3b7db158013ad35d368b95626c884775542a156/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolSpecBuilder.java#L313-L348

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionUtils.java
Comment (siladu) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionUtils.java]: Refactored this out in https://github.com/hyperledger/besu/pull/4743/commits/698849ec77c38079d788b23b6ab846b8d592643e
It did have a small knock-on effect to TransitionCoordinator and another caller of the TransitionBesuControllerBuilder constructor without the mergeContext, so arguably has leaked PostMergeContext.get() to an abstraction level too high, but I don't think it's a big deal.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java]: done

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java:30]: committed something for this including a new PrivacySupportingProtocolSchedule

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java]: I've removed it and renamed ProtocolSpecAdapter variable to `blockNumberOrTimestamp` <-- I'm dubious about this though because it implies you can mix number and timestamp in the same modifiers Map. I'm not sure that could happen with the current implementation of TimestampSchedule. 
I'm also not sure if it's actually a problem to mix them either, would technically work as long as blocks don't follow timestamps.

An alternative approach would be to create a type hierarchy to discriminate between blockNumber and timestamp, but I've resisted this for now because the SpecAdapters won't be needed for mainnet/testnets with #4788
 
Another approach would be to leave ProtocolSpecAdapters as blockNumber only don't attempt to support other usecases such as Clique.

Comment (jframe) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampProtocolSpecAdapters.java]: Think it's ok to use `blockNumberOrTimestamp` for now.

Would prefer that it contained only one type as doesn't handle ordering between block numbers and timestamps unless timestamps strictly follow block numbers. Using generics and type hierarchy to enforce this would be nice so we could have `ProtocolSpecAdapters<Timestamp>` for example. But not sure it's worth the effort either as ProtocolSpecAdapters will go away soon for the timestamp case once #4788 is done.

File: besu/src/main/java/org/hyperledger/besu/controller/TransitionBesuControllerBuilder.java
Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/controller/TransitionBesuControllerBuilder.java:189]: Have added a comment to #4789 description about this

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java:34]: OK, I've implemented it here: https://github.com/gezero/besu/pull/1

It was easy enough...question is though, is it worth the extra complexity for the future code reader?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java]: should this be >= 0L ?  for a merge-at-genesis ?  https://github.com/hyperledger/besu/pull/4792 is allowing for merge on genesis block, with no transition schedule (which presumably would be extended to include a timestamp protocol schedule at time 0 also)

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java]: this could/should be pulled from all forks if we allow timestamps =1 (or 0).  Since ethereum started well after epoch time.  It is a corner case, but test networks might make some interesting assumptions for their genesis

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java]: this is a more verbose version of the same thing?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java]: seems we are doing a lot of concatenating.  we should just maintain a composed list of forks by number and timestamp

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java
Comment (garyschulte) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java:30]: disagree about the privacy stuff.  I think we should be excising it in favor of mainnet compatibility.  $0.02

**Comment by siladu** [comment]
@gezero your last force pushed removed some commits, so I've just forced pushed again to readd them. I've also removed the ForkId commit from this PR as it's nearly ready to be merged and didn't want to block it on ForkId. Your commit is now on https://github.com/hyperledger/besu/pull/4815 and @jframe is continuing the implementation

**Review by garyschulte — COMMENTED** [review]
phew :sweat_smile: 

Some feedback is opinionated, so I will defer to you @jframe @gezero

File: config/src/main/java/org/hyperledger/besu/config/JsonGenesisConfigOptions.java
Comment (siladu) [reply config/src/main/java/org/hyperledger/besu/config/JsonGenesisConfigOptions.java:283]: I have asked on Discord, still waiting for a response, but I will likely change to shanghaiTime at least. Maybe create an alias for shanghaiBlock if necessary

File: config/src/test/java/org/hyperledger/besu/config/GenesisConfigFileTest.java
Comment (siladu) [reply config/src/test/java/org/hyperledger/besu/config/GenesisConfigFileTest.java]: Sorry, ForkId longer part of this PR, it had no test coverage so it's moved to https://github.com/hyperledger/besu/pull/4815/commits/8b5e3ed1092bdd2b58851ba123976268a41044bf

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/MergeProtocolSchedule.java
Comment (siladu) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/MergeProtocolSchedule.java:75]: I think there's a discussion to be had around whether or not we support timestamps for non-mainnet. I've tried to leave it open for both. Extra code to support it has been minimal and I don't think it's caused any problems so far, other than more 'noise' in the code...maybe warrants a manual test with a private network.

I don't feel too strongly either way, support can easily be added back in later if necessary.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/ProcessableBlockHeader.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/ProcessableBlockHeader.java:163]: > block short encoding
Not sure what you mean by this...I believe this is intended for a log message only. 

Happy to go with `time`, `ts` is a bit obscure IMO

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockHeader.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockHeader.java:337]: It might be unnecessary. I only included timestamp in the other one because it didn't have access to getHash and thought something was better than nothing. In reality, this impl will be used instead of the ProcessableBlockHeader impl, and I'm not sure what value adding timestamp would bring above the hash.

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestProtocolSchedules.java
Comment (siladu) [reply ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestProtocolSchedules.java:76]: Should be able to use Shandong instead of Shanghai still. Shandong is being removed anyway. 
Once this PR is in, we can start migrating Shandong guards -> Shanghai

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java]: That's one for https://github.com/hyperledger/besu/pull/4815/commits/8b5e3ed1092bdd2b58851ba123976268a41044bf cc @jframe

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java]: Another for https://github.com/hyperledger/besu/commit/8b5e3ed1092bdd2b58851ba123976268a41044bf cc @jframe

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/DefaultTimestampSchedule.java:30]: I think we should discuss with hyperledger community whether timestamp support is desirable

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/MergeProtocolSchedule.java
Comment (jframe) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/MergeProtocolSchedule.java:75]: I was taking the position that if we can support a feature for private networks then by default we do because I think that's the approach we have been taking. But there probably isn't a lot of value in supporting this for private networks. So if the code is simpler by not supporting private networks then we could remove the timestamp support for private networks.

Comment (jframe) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/MergeProtocolSchedule.java:75]: For the `MergeProtocolSchedule` though I can't think of my we would need to have the privacy parameters. For IBFT or QBFT we use a different protocol schedule.

Don't think this is necessary to block the PR though, so could clean this up later.

**Review by jframe — APPROVED** [review]

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/TimestampScheduleBuilder.java]: After speaking to Jason, decide to add the refactor into this PR.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/ProcessableBlockHeader.java
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/ProcessableBlockHeader.java:163]: on second thoughts, timestamp is the name of the field on the block so think even 'time' would confuse things here.

File: config/src/main/java/org/hyperledger/besu/config/JsonGenesisConfigOptions.java
Comment (siladu) [reply config/src/main/java/org/hyperledger/besu/config/JsonGenesisConfigOptions.java]: For now, I've gone with shanghaiTime as I think that's more likely than shanghaiTimestamp

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/TransitionProtocolScheduleTest.java
Comment (siladu) [reply consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/TransitionProtocolScheduleTest.java]: Added a test and updated the name to say pre or post

**Comment by siladu** [comment]
@bobsummerwill FYI: https://github.com/ethereum/EIPs/pull/6122

**Pull request merged**
